app/controllers/bulk_ingest_controller.rb

```diff
@@ -90,8 +60,124 @@ def source_metadata_id_from_path
       base_path if valid_remote_identifier?(base_path)
     end
 
+    def browse_everything_params
+      return {} unless params.key?("browse_everything")
+      @browse_everything_params ||= params["browse_everything"]
+    end
+
+    def browse_everything_uploads
+      return [] unless browse_everything_params.key?("uploads")
+      @browse_everything_uploads ||= browse_everything_params["uploads"]
+    end
+
+    def find_upload(upload_id)
+      uploads = BrowseEverything::Upload.find_by(uuid: upload_id)
+      uploads.first
+    end
+
+    def new_uploads
+      @new_uploads ||= begin
+                         browse_everything_uploads.map do |upload_id|
+                           find_upload(upload_id)
+                         end
+                       end
+    end
+
+    def ingest_multi_volume_works(change_set_persister)
+      new_uploads.each do |upload|
+        file_tree = {}
+        upload.files.each do |upload_file|
+          new_pending_upload = PendingUpload.new(
+            id: SecureRandom.uuid,
+            upload_id: upload.id,
+            upload_file_id: upload_file.id
+          )
+
+          if new_pending_upload.in_container?
+            pending_uploads = file_tree[upload_file.container_id] || []
+            file_tree[upload_file.container_id] = pending_uploads + [new_pending_upload]
+          end
+        end
+
+        directory_tree = {}
+        parent_containers = []
+        upload.containers.each do |container|
+          # Are there files for this container?
+          if file_tree.key?(container.id)
+            # Create the volume work
+            children = file_tree[container.id]
+            volume_name = container.name
+            member_change_set = build_change_set(title: volume_name, pending_uploads: children, files: children)
+
+            persisted = change_set_persister.save(change_set: member_change_set)
+
+            parent_id = container.parent_id
+            if parent_id
+              members = directory_tree[parent_id] || []
+              directory_tree[parent_id] = members + [persisted]
+            end
+          else
+            parent_containers << container
+          end
+        end
+
+        parent_containers.each do |container|
+          # If not, create a parent work
+          members = directory_tree[container.id] || []
+          member_ids = members.map(&:id)
+          parent_name = container.name
+
+          parent_change_set = build_change_set(title: parent_name, member_ids: member_ids)
+          change_set_persister.save(change_set: parent_change_set)
+        end
+      end
+    end
+
+    def ingest_works(change_set_persister)
+      new_uploads.each do |upload|
```

Comment:
This seems like the same method as in `ingest_multi_volume_works`, which points again to maybe some extraction we can do. This controller's real big too.